### PR TITLE
Grpc14

### PR DESCRIPTION
**Pull request by GuillaumeCisco**
grpc 14 error handling with retries.

https://forum.zeebe.io/t/is-grpc-a-point-of-failure-when-embedding-the-workflow-in-a-program/455/12

### CODE REVIEW

**Comment by dexhunter** [comment]
The build failure was caused by flake8 regulations.

**Comment by GuillaumeCisco** [comment]
Yeah, I know.
This is a work in progress.
Do you have any thoughts about this PR.

After a lot of thinking, I came to the conclusion the fabric sdk should not include a method line invoke as it can be handled totally differently regarding the endorsement policy...

**Comment by dexhunter** [comment]
> fabric sdk should not include a method line invoke as it can be handled totally differently regarding the endorsement policy...

Would you like to elaborate on that a bit more? Thanks

**Comment by GuillaumeCisco** [comment]
The way we handle the invocation could be very different between apps.

With a basic `OR` policy, we should not even retry on failed proposal responses and simply feed the orderer when we get at least one good proposal.
It can be the same for `MAJORITY` policy, but we could be sure at least the majority of proposals got status 200.
Regarding `AND` policy or slightly more complex policy, using the retry strategy could be great!

Fabric sdk node does not implement this method but show way to create it in its tutorial.

Maybe we could add a warning inside the chaincode invoke method or our tutorial telling the user he should it itself.
We could even create a high level fabric sdk python project showing how to deal with invoke on different kind of endorsment policies. Users could use them or simply be inspired by them for creating their own.

I hope I'm explicit enough, do not hesitate for more clarity.

File: hfc/fabric/client.py [hunk 1]
```diff
@@ -1632,17 +1638,52 @@ def cc_event(self, cc_event, block_number, tx_id, tx_status):
...
+            if grpc_broker_unavailable_retry:
+                retry = 0
+                ok = False
+
+                # recreate responses from unreachable peers
+                failed_target_peers = list(itertools.compress(target_peers, failed_res))
+
+                while retry < grpc_broker_unavailable_retry:
+                    _logger.debug(f'Retrying getting proposal responses from peers: {[x.name for x in failed_target_peers]}, retry: {retry}')
+                    print(f'Retrying getting proposal responses from peers: {[x.name for x in failed_target_peers]}, retry: {retry}')
```
Comment (Kelvin-M) [inline hfc/fabric/client.py]: Why do you use a logger and a print to display the same information ?

**Comment by Kelvin-M** [comment]
Thank you for this work!
As we all know, the chaincode invoke can be a complicated function and other sdk let this implementation to the final user.

Retrying on fail broker is a good strategy as we may have network interruption (not the only reason for the grpc 14 issue, but the most probable).

I think, even if all brokers does not respond, we must send the endorsed proposal to the orderer and not raise an exception (https://github.com/hyperledger/fabric-sdk-py/pull/18/files#diff-d627d0a212052fa058ddb5ef73df231eR1669-R1671)
If the endorsed proposal is good (correctly endorsed along to the chaincode policy) the orderer will confirm it. In the opposite case, the orderer will trigger a error which can be return to the user.

We may also have an option to tell this function if we send the proposal even if all brokers do not endorsed the proposal.

File: hfc/fabric/client.py [hunk 2]
```diff
@@ -1619,17 +1623,57 @@ def cc_event(self, cc_event, block_number, tx_id, tx_status):
...
+
+                    retry += 1
+                    # TODO should we use a backoff?
+                    _logger.debug(f'Retry in {grpc_broker_unavailable_retry_delay}ms')
+                    sleep(grpc_broker_unavailable_retry_delay / 1000)  # milliseconds
+
+                if len(failed_target_peers) > 0:
+                    # TODO should we really raise or send to the orderer without issue?
+                    raise Exception(f'Could not reach peer grpc broker {[x.name for x in failed_target_peers]}'
+                                    f' even after {grpc_broker_unavailable_retry} retries.')
```
Comment (Kelvin-M) [inline hfc/fabric/client.py]: Can you put a option to choose if we want to send or not the proposal even if endorsement with some brokers have failed ?

**Comment by Kelvin-M** [comment]
@GuillaumeCisco @dexhunter Where are we on this one ?
I could add the feature from my comment, but are you ok with :) ?

**Comment by dexhunter** [comment]
> @GuillaumeCisco @dexhunter Where are we on this one ?
> I could add the feature from my comment, but are you ok with :) ?

I am okay, you can update with a new PR. Thanks!

**Pull request merged**